### PR TITLE
KFSPTS-25263 create base batch job to upload suppliers to Jaggaer

### PR DESCRIPTION
**Pull request by jhulslander**
Please note there is a related non-prod SQL PR.  Don't merge this till both PRs are ready.
There will be a follow up user story to actually create the CSV file, and to make use of the FIPS to ISO mapping service.

I noticed JaggaerUploadDaoJdbc does include some static final variable that I think exist in base code.  I think this PR is ready for review, but please note I want to look into adjusting those variables

### CODE REVIEW

**Review by ajd299 — CHANGES_REQUESTED** [review]
Please see my suggested changes, most of them are just typos

**Comment by jhulslander** [comment]
retest this please

File: src/main/java/edu/cornell/kfs/module/purap/batch/JaggaerGenerateContractPartyCsvStep.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        return getParameterValueString(CUPurapParameterConstants.JAGGAER_UPLOAD_PO_DATE);
+    }
+    
+    protected String findVendorDate() {
+        return getParameterValueString(CUPurapParameterConstants.JAGGAER_UPLOAD_VENDOR_DATE);
+    }
+    
+    protected String getParameterValueString(String paramterName) {
+        return parameterService.getParameterValueAsString(this.getClass(), paramterName);
+    }
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/JaggaerGenerateContractPartyCsvStep.java:66]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                dto.setLegalStructure(JaggaerLegalStructure.findJaggaerLegalStructureByKFSOwnershipCode(resultSet.getString(FIELD_NAMES.VNDR_OWNR_CD)));
+                dto.setTaxIDType(StringUtils.EMPTY);
+                dto.setTaxIdentificationNumber(StringUtils.EMPTY);
+                dto.setVATRegistrationNumber(StringUtils.EMPTY);
+                dto.setWebsiteURL(resultSet.getString(FIELD_NAMES.VNDR_URL_ADDR));
+                return dto;
+            };
+            return this.getJdbcTemplate().query(buildVendorSql(processingMode, processingDate), rowMapper);
+        } catch (Exception e) {
+            LOG.error("findJaggaerContractParty, had an error finding jaggaer contract partiess: ", e);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:51]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            sb.append("AND VD.LAST_UPDT_TS > TO_DATE('").append(processingDate).append("', 'YYYY-MM-DD') ");
+        }
+        sb.append("ORDER BY VA.VNDR_HDR_GNRTD_ID, VA.VNDR_DTL_ASND_ID");
+        
+        String sql = sb.toString();
+        LOG.info("buildVendorAddressSql, SQL: " + sql);
+        return sql;
+    }
+    
+    private String buildPurchasOrderLimitSubQuery(String processingDate) {
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:136]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        sb.append("SELECT VNDR_HDR_GNRTD_ID, COUNT(1) ");
+        sb.append("FROM KFS.PUR_PO_T ");
+        sb.append("WHERE PO_LST_TRNS_DT > TO_DATE('").append(processingDate).append("', 'YYYY-MM-DD') ");
+        sb.append("GROUP BY VNDR_HDR_GNRTD_ID ");
+        sb.append("HAVING COUNT(1) > 1)");
+        return sb.toString();
+    }
+    
+    private String buildVendorNumber(String vendorHeaderId, String venderDetailId) {
+        return vendorHeaderId + KFSConstants.DASH + venderDetailId;
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/dataaccess/impl/JaggaerUploadDaoJdbc.java:147]: ok

File: src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            }
+        }
+        return uploadDtos;
+    }
+    
+    protected List<JaggaerContractAddressUploadDto> findVendorAddresses(List<JaggaerContractAddressUploadDto> allAddresses, String erpNymber) {
+        return allAddresses.stream()
+                .filter(address -> StringUtils.equalsIgnoreCase(address.getERPNumber(), erpNymber))
+                .collect(Collectors.toCollection(ArrayList::new));
+    }
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImpl.java:50]: ok

File: src/main/java/edu/cornell/kfs/module/purap/businessobject/lookup/JaggaerContractPartyUploadDto.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+    public void setWebsiteURL(String websiteURL) {
+        this.websiteURL = websiteURL;
+    }
+    
+    @Override
+    public String toString() {
+        ReflectionToStringBuilder builder = new ReflectionToStringBuilder(this, ToStringStyle.NO_CLASS_NAME_STYLE);
+        builder.setExcludeFieldNames("taxIdentificationNumber");
+        builder.append("taxIdentificationNumber", StringUtils.isBlank(taxIdentificationNumber) ? StringUtils.EMPTY : "restrctived tax id number");
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/module/purap/businessobject/lookup/JaggaerContractPartyUploadDto.java:127]: switched to Restricted

File: src/main/java/edu/cornell/kfs/vnd/CUVendorConstants.java [hunk 7]
```diff
@@ -19,6 +19,8 @@
 import java.util.LinkedHashMap;
 import java.util.Map;
 
+import edu.cornell.kfs.vnd.CUVendorConstants.FIELD_NAMES;
+
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/vnd/CUVendorConstants.java:23]: nope, not needed, thanks

File: src/main/java/edu/cornell/kfs/vnd/CUVendorConstants.java [hunk 8]
```diff
@@ -79,4 +81,23 @@ public static class CUAddressTypes {
     }
     
     public static final int MAX_VENDOR_CONTACT_NAME_LENGTH = 45;
+    
+    public final class FIELD_NAMES {
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/vnd/CUVendorConstants.java:85]: yeah, it should be static

File: src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    @Test
+    void testGetJaggerContractsDto() {
+        jaggaerGenerateContractPartyCsvServiceImpl.setJaggaerUploadDao(buildMockJaggaerUploadDao());
+        List<JaggaerContractUploadBaseDto> jaggaerUploadDtos = jaggaerGenerateContractPartyCsvServiceImpl.getJaggerContractsDto(JaggaerContractUploadProcessingMode.VENDOR, PROCESS_DATE);
+        assertEquals(4, jaggaerUploadDtos.size());
+        
+        JaggaerContractPartyUploadDto actualFirstELement = (JaggaerContractPartyUploadDto) jaggaerUploadDtos.get(0);
+        assertEquals(buildVendor1Dto(), actualFirstELement, "The first element shound be vendor 1");
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:161]: ok

File: src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        JaggaerContractAddressUploadDto expectedAddress1 = buildVendor1Address1Dto();
+        expectedAddress1.setName(actualFirstELement.getContractPartyName());
+        assertEquals(expectedAddress1, actualSecondElement, "The second element should be vendor 1 address 1");
+        
+        JaggaerContractAddressUploadDto actualThirdElement = (JaggaerContractAddressUploadDto) jaggaerUploadDtos.get(2);
+        JaggaerContractAddressUploadDto expectedAddress2 = buildVendor1Address2Dto();
+        expectedAddress2.setName(actualFirstELement.getContractPartyName());
+        assertEquals(expectedAddress2, actualThirdElement, "The third element should be vendor 1 address 2");
+        
+        JaggaerContractPartyUploadDto actualFourthELement = (JaggaerContractPartyUploadDto) jaggaerUploadDtos.get(3);
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:173]: ok

File: src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        assertEquals(buildVendor2WithTaxIdDto(), actualFourthELement, "THe fourth elemeent should be vendor 2");
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:174]: ok

File: src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    @ParameterizedTest
+    @MethodSource("jaggaerContractPartyUploadDtoToStringParameters")
+    void testJaggaerContractPartyUploadDtoToString(JaggaerContractPartyUploadDto contractPartyDto, String searchString) {
+        String actualToString = contractPartyDto.toString();
+        assertTrue(StringUtils.contains(actualToString, searchString));
+    }
+    
+    static Stream<Arguments> jaggaerContractPartyUploadDtoToStringParameters() {
+        return Stream.of(
+                Arguments.of(buildVendor2WithTaxIdDto(), "taxIdentificationNumber=restrctived tax id number"),
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/test/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerGenerateContractPartyCsvServiceImplTest.java:186]: restricted

**Comment by jhulslander** [comment]
I've made the changes suggested, I also discovered a problem with the SQL that I have corrected, in doing so I simplified and reduced some duplication in how the SQL is generated in JaggaerUploadDaoJdbc.

This is ready for review again

**Review by ajd299 — APPROVED** [review]
Code looks good, merging

**Pull request merged**